### DISCUSSION TITLE
Warm Start with dual variables

### DISCUSSION THREAD

**Discussion by GTRay** [coin-or/Ipopt#605] [Q&A] [answered]
Hi, I am using Ipopt to solve a NLP problem, and would like to use the solution from previous cycle to warm start the next cycle. I did some research and it looks like I just need to do `app->Options()->SetStringValue("warm_start_init_point","yes");`. However, the boolean input `init_z` and `init_lambda` to `MyNLP::get_starting_point()` are still both false, so I don't think the dual variables from previous run are fed into the solver. There are some previous discussion I found in https://list.coin-or.org/pipermail/ipopt/2016-May/004221.html, but still not sure if I set up the warm start correctly. So what is the right way to warm start the solver? Can I have some help? Thanks.

**Comment by svigerske** [comment, marked as answer]
Setting `warm_start_init_point` to `yes` should be sufficient.
If I add an ipopt.opt with `warm_start_init_point yes` and run the hs071_cpp example, I get
```
lt-hs071_cpp: hs071_nlp.cpp:120: virtual bool HS071_NLP::get_starting_point(Ipopt::Index, bool, Ipopt::Number*, bool, Ipopt::Number*, Ipopt::Number*, Ipopt::Index, bool, Ipopt::Number*): Assertion `init_z == false' failed.
```
So `init_z` seems to be true.

Please check your code again.

**Reply by GTRay** [reply]
> Setting `warm_start_init_point` to `yes` should be sufficient. If I add an ipopt.opt with `warm_start_init_point yes` and run the hs071_cpp example, I get
> 
> ```
> lt-hs071_cpp: hs071_nlp.cpp:120: virtual bool HS071_NLP::get_starting_point(Ipopt::Index, bool, Ipopt::Number*, bool, Ipopt::Number*, Ipopt::Number*, Ipopt::Index, bool, Ipopt::Number*): Assertion `init_z == false' failed.
> ```
> 
> So `init_z` seems to be true.
> 
> Please check your code again.

Thanks for the reply. Stefan. I have tried the example you mentioned on `stable/3.14`. There are a few interesting things:
1. `assert` doesn't seem to have any effect
2. `get_starting_point()` seems to get called twice. First time with
```
init_x true
init_z false
init_lambda false
```
then second time with
```
init_x true
init_z true
init_lambda true
```
I guess it kind of makes sense, because warm start would need the previous solution to boot up, so the first run has `init_z` and `init_lambda` being false. So what is the expected behavior when setting `warm_start_init_point yes`? Is there a way to skip the first run when `init_z` and `init_lambda` are false?

**Reply by svigerske** [reply]
Asserts only show up when Ipopt is build with debugging enabled (which isn't default).

OK, now I remember: The first call is when Ipopt computes the problem scaling. For this, it needs to evaluate all functions, which it does at the starting point.
The second call is then to determine the actual starting point for the solve.

The only way to skip the first call is to disable scaling or provide scaling parameters by yourself.
Best would be to just not provide starting values for z and lambda when they are not requested.

**Reply by GTRay** [reply]
> Asserts only show up when Ipopt is build with debugging enabled (which isn't default).
> 
> OK, now I remember: The first call is when Ipopt computes the problem scaling. For this, it needs to evaluate all functions, which it does at the starting point. The second call is then to determine the actual starting point for the solve.
> 
> The only way to skip the first call is to disable scaling or provide scaling parameters by yourself. Best would be to just not provide starting values for z and lambda when they are not requested.

Ah, gotcha, that makes a lot of sense. Really appreciate it. Stefan.
